fix(orders): fix cancel order race condition using time-based cache

shamardy opened the pull request:
This PR addresses the issue where cancelled orders can sometimes remain in the orderbook due to P2P message caching and out-of-order message delivery.

Changes:

- Added a time-based cache to orderbook to track recently cancelled orders
- Modified order processing logic to check the cancellation cache before processing create maker order messages.
- Set cache expiration time to 2 minutes, exceeding the 1-minute gossipsub cache to ensure coverage of delayed messages.
- Added `maker_order_cancelled_p2p_notify` to a location where it was previously missing, this needs careful review. Old PR that added the maker order deletion without notifying https://github.com/KomodoPlatform/komodo-defi-framework/pull/1178 and the reason for the old PR https://github.com/KomodoPlatform/komodo-defi-framework/issues/1173#issuecomment-1000813076

The solution works as follows:

1. When an order is cancelled, its UUID is added to the cancellation cache with a 2 minutes expiration time.
2. Incoming create maker order messages are checked against the cache before processing.
3. If a matching UUID is found in the cache and the pubkey is the maker pubkey, the create maker order message is ignored.
4. Cache entries expire after 2 minutes to reduce memory usage.

This approach provides a simple fix for the immediate issue while avoiding major changes to the P2P layer or gossipsub protocol. It prevents cancelled orders from reappearing in the orderbook due to delayed or out-of-order message delivery.

Next steps:

- Monitor for any new issues related to P2P message ordering.
- Consider more comprehensive solutions for ensuring message ordering if needed in the future.

potentially fixes https://github.com/KomodoPlatform/komodo-defi-framework/issues/1726

mariocynicys commented on mm2src/mm2_main/tests/mm2_tests/mm2_tests_inner.rs:
```diff
@@ -1674,9 +1674,10 @@ fn test_cancel_all_orders() {
-    let pause = 3;
-    log!("Waiting ({} seconds) for Bob to cancel the order…", pause);
-    thread::sleep(Duration::from_secs(pause));
+    block_on(mm_alice.wait_for_log(5., |log| {
+        log.contains(&format!("Maker order {} was recently cancelled, ignoring", uuid))
+    }))
+    .unwrap();
```
electrum PR fails here. passes in `test_cancel_order` though.

shamardy replied:
Try the old test without this log, it will work for sure. Sometimes the messages arrive in the right order it seems.

shamardy replied:
Will fix this to not rely on the log in case messages arrive in the right order (which happens sometimes of course)

shamardy replied:
Something interesting happened @mariocynicys , `test_cancel_all_orders` failed here after I cherry picked the changes to expirable map from your branch https://github.com/KomodoPlatform/komodo-defi-framework/actions/runs/11100978215/job/30837774256?pr=2232#step:7:1547 . I will solve it by this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#discussion_r1780529737 but we found the changes that makes it fail there (not due to old out of order messages reason but I mean this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#discussion_r1779464754)

shamardy replied:
> idk i think macos used to fail it sometimes in the past, but can't verify since logs get purged by github. this failure isn't consistent though since it only happened on macos.
and the test still fails after reverting the expirable map changes :(

No problem, will fix it by not relying on the log and everything should be fine.

shamardy replied:
This https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/87c53f81ed7cc727f6af4715a6289bd82d5a8bc1 should work on the electrums PR, the test also covers the case of out of order messages when it happens, not in every run though :)

shamardy replied:
macOS still unstable and fails sometimes. I think it's another issue and the test is fine.

mariocynicys reviewed (commented):
took a quick look and tried the fix on the electrum connections branch (since it exemplifies the issue 1000x for some reason).

I fear though that nodes don't receive messages at the same time and we have some node getting the message AFTER (because some other node received the cancellation late enough that it's still in its gossibpub cache) the cancellation cache timed out and it would consider it a new order creation.

can't we enforce in-order-delivery somehow on p2p layer? or at leas t only causal delivery.

shamardy commented on mm2src/common/expirable_map.rs:
```diff
@@ -47,6 +47,7 @@ impl<K: Eq + Hash, V> ExpirableMap<K, V> {
     /// If a value already exists for the given key, it will be updated and then
     /// the old one will be returned.
     pub fn insert(&mut self, k: K, v: V, exp: Duration) -> Option<V> {
+        self.clear_expired_entries();
```
I might do some optimization to this https://github.com/KomodoPlatform/komodo-defi-framework/commit/62d26c7b59f2e47f6e50b7b1ecb6d66b8dcdf5ab. Will push the cherry-pick first then do the optimizations in another commit so that you can review them.

shamardy replied:
> Only optimization I can find currently is to use retain in clear_expired_entries instead of for (exp, key) in keys_to_remove { loop

Nevermind, your current implementation is better :)

shamardy replied:
Done here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/1c38da77a11be4eff685cf2e4f2f782d3021a94e

onur-ozkan replied:
`remove` function calls `clear_expired_entries` and I think it's not ideal. The entire implementation relies on manually managed expired entries. If we want to handle this automatically, we should do it in every function not just one. Currently, if I’ve already called `clear_expired_entries`, using the `remove` function adds unnecessary overhead. I’m also wondering if we can avoid `Clone`. I will look into that.

onur-ozkan replied:
Is `cloned` necessary?

shamardy replied:
> Is `cloned` necessary?

Found a better alternative that doesn't use cloned
```rust
    pub fn clear_expired_entries_impl1(&mut self) {
        let now = Instant::now();

        while let Some((exp, key)) = self.expiries.pop_first() {
            if exp > now {
                self.expiries.insert(exp, key);
                break;
            }
            self.map.remove(&key);
        }
    }
```
<img width="838" alt="Screenshot 2024-09-30 at 11 35 44 AM" src="https://github.com/user-attachments/assets/74f0a546-1c25-4700-9dc1-1d48a4685922">

shamardy replied:
> I’m also wondering if we can avoid Clone. I will look into that.

No need to check it yourself, I will check it now.

shamardy replied:
Implemented this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#discussion_r1780664590 here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/da884fcb87a831716c524e9b317846d942b7289a

onur-ozkan replied:
I think we should either avoid calling `clear_expired_entries` anywhere in `ExpirableMap` or call it consistently throughout and update the documentation to reflect this (as it currently stands as manually managed expired entries).

The downside of calling `clear_expired_entries` within the binding functions is that when you insert/remove/get entries several times, `clear_expired_entries` will be triggered that many times within milliseconds. This is why it was designed in a manual way initially by leaving expiration management responsibility to the caller (e.g., I can clear the expired entries once and then insert/get/remove multiple keys).

shamardy replied:
> I’m also wondering if we can avoid Clone. I will look into that.

Can't remove it without changing the current `ExpirableMap` structure, we can use `Arc` but it doesn't make sense, I think cloning once on insert is fine unless you have other ideas @onur-ozkan.

shamardy replied:
> We don't really use get anywhere but I don't mind. The only thing is that it will have to be mutable func.
Also i think we can call clear_expired_entries on remove just fine. It's cheap and why wouldn't we.

No need to then, we can just check expiry on `get` and return `None` if expired.

onur-ozkan replied:
My concern is that we are semi-managing the expirations, which is confusing on use (we have to check which function are clearinging expired entries and which functions are not clearing it..). How about making `clear_expired_entries` private function, and using it internally everywhere in `ExpirableMap` implementation? It should be very clear that `ExpirableMap` is either controlling expired entries, or not doing it at all.

> Can't remove it without changing the current ExpirableMap structure, we can use Arc but it doesn't make sense, I think cloning once on insert is fine unless you have other ideas @onur-ozkan.

Yeah, it doesn't make sense to use `Arc`. What about using `Copy` instead of `Clone`? This should enforce callers to use `Copy`able (which are cheaper) types for keys.

shamardy replied:
> Also i think we can call clear_expired_entries on remove just fine. It's cheap and why wouldn't we.

There is no need for `clear_expired_entries` on remove, the purpose of `clear_expired_entries` is to reduce the memory and we can do it on insert only.

shamardy replied:
> Yeah, it doesn't make sense to use Arc. What about using Copy instead of Clone? This should enforce callers to use Copyable (which are cheaper) types for keys.

Will give copy a try.

shamardy replied:
To summarize:

1 - Will make `clear_expired_entries` private
2 - Use it only on `insert`
3 - check expiry on `get`, not sure but we shouldn't return an expired entry in anyway.

onur-ozkan replied:
> > Also i think we can call clear_expired_entries on remove just fine. It's cheap and why wouldn't we.
> 
> There is no need for `clear_expired_entries` on remove, the purpose of `clear_expired_entries` is to reduce the memory and we can do it on insert only.

I agree for `remove`, but if we doing it for `insert`, we should do it for `get` too IMO. And if we do it for `insert` and `get`, `clear_expired_entries` can be private function.

shamardy replied:
to do it on get, get will need a mutable self.

onur-ozkan replied:
Oooh, yeah.. That's not good.. I don't know what to suggest at this point :laughing:. As I said, a consistent interface is the concern for me. If it's not going to happen in every function (except remove as it doesn't make any sense), it should be handler from the caller side IMO as it's much clear to follow.

shamardy replied:
Did this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#discussion_r1780828764 here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/5762d13c602d72c25ca2ae060c4402366235cebb

shamardy replied:
Will try to replace clone with copy now.

shamardy replied:
> Will try to replace clone with copy now.

Done here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/f2ef6d3a132d1ce52ea6ed3b085314a757c7bb6f

mariocynicys replied:
> There is no need for clear_expired_entries on remove, the purpose of clear_expired_entries is to reduce the memory and we can do it on insert only.

Not just to reduce the memory, but also to enforce correctness. But if we are changing the API to tell the caller that `get` is the clearing version then it's OK.

mariocynicys replied:
> What about using Copy instead of Clone? This should enforce callers to use Copyable (which are cheaper) types for keys.

Haha, i started with the same approach but chose clone eventually cuz tests used strings (non-copyable). Agree, this should be the way though.

shamardy commented:
Received this message from @cipig 

> i run 2 nodes with this PR, the last commits are missing. I noticed that those 2 nodes showed this
> ```
> 30 22:00:01, mm2_p2p::behaviours::atomicdex:324] ERROR notify_on_adex_event error send failed because channel is full
> 30 22:00:01, mm2_p2p::behaviours::atomicdex:324] ERROR notify_on_adex_event error send failed because channel is full
> 30 22:00:01, mm2_p2p::behaviours::atomicdex:324] ERROR notify_on_adex_event error send failed because channel is full
> 30 22:00:05, libp2p_gossipsub::behaviour:3415] INFO HEARTBEAT: relays low. Contains: 0 needs: 2
> ```
> at the same time, while the other nodes didn't
> question is if the changes increase the probability of this to happen

Will look into it after fixing the review comments

shamardy commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -3314,6 +3375,7 @@ pub async fn lp_ordermatch_loop(ctx: MmArc) {
                 // This checks that the order hasn't been removed by another process
                 if let Some(order_mutex) = removed_order_mutex {
                     let order = order_mutex.lock().await;
+                    maker_order_cancelled_p2p_notify(ctx.clone(), &order);
```
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/69fdfe985710b13f204ebc86e525be30b9eb7b2e

shamardy commented:
About this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#issuecomment-2385098962 , only reason for delay is holding the orderbook lock for some time, I can move the new map to `OrdermatchContext` instead with its own lock to reduce the delay.

onur-ozkan commented:
> Received this message from @cipig
> 
> > i run 2 nodes with this PR, the last commits are missing. I noticed that those 2 nodes showed this
> > ```
> > 30 22:00:01, mm2_p2p::behaviours::atomicdex:324] ERROR notify_on_adex_event error send failed because channel is full
> > 30 22:00:01, mm2_p2p::behaviours::atomicdex:324] ERROR notify_on_adex_event error send failed because channel is full
> > 30 22:00:01, mm2_p2p::behaviours::atomicdex:324] ERROR notify_on_adex_event error send failed because channel is full
> > 30 22:00:05, libp2p_gossipsub::behaviour:3415] INFO HEARTBEAT: relays low. Contains: 0 needs: 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > at the same time, while the other nodes didn't
> > question is if the changes increase the probability of this to happen
> 
> Will look into it after fixing the review comments

How is this PR related with those logs :thinking:

shamardy commented:
> About this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#issuecomment-2385098962 , only reason for delay is holding the orderbook lock for some time, I can move the new recently cancelled map to OrdermatchContext instead with its own lock to reduce the delay.

> How is this PR related with those logs 🤔

I can see that `process_msg` is async, so you are right, it shouldn't cause messages buffer to be full. But I should add this optimization nonetheless.

shamardy commented:
> I can move the new recently cancelled map to OrdermatchContext instead with its own lock to reduce the delay.

Done here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/14f7f340f15f7ee7d4433839677e1d92a90c154a

shamardy commented:
@onur-ozkan @mariocynicys all issues raised in review comments should have been fixed. Please give this another look when you have the time.

onur-ozkan commented on mm2src/common/expirable_map.rs:
```diff
@@ -19,48 +19,80 @@ impl<V> ExpirableEntry<V> {
+    fn clear_expired_entries(&mut self) {
+        let now = Instant::now();
+
+        while let Some((exp, key)) = self.expiries.pop_first() {
+            if exp > now {
+                self.expiries.insert(exp, key);
```
I think `first_key_value` should improve the performance here. For non-expired entries (which means for a lot of entries) this has overhead with removing and inserting back the same value.

onur-ozkan replied:
> With pop_first removing and inserting back will only be done once if the first entry is a non-expired one.

It should happen more than *once* as there can be a lot of non-expired records (non-expired entries will be more than the expired ones at most times).

> If we use first_key_value we will have to remove the entry after that from expiries BTreeMap which adds a lot of overhead.

Instead of removing and inserting back unconditionally, iterating over references then removing when needed should be much quicker I think. Did you measure the performance difference between these two?

shamardy replied:
> It should happen more than once as there can be a lot of non-expired records (non-expired entries will be more than the expired ones at most times).

But it's ordered so when the first case happens we end the while loop.

> Instead of removing and inserting back unconditionally, iterating over references then removing when needed should be much quicker I think. Did you measure the performance difference between these two?

I can measure it if needed :)

onur-ozkan replied:
> But it's ordered so when the first case happens we end the while loop.

I missed that part, yeah. I think we should add some explanation comments inside this function. It will look very confusing for other people.

shamardy replied:
> I think we should add some explanation comments inside this function. It will look very confusing for other people.

Done.

onur-ozkan reviewed (commented):
Last notes from my side:

shamardy commented:
> only reason for delay is holding the orderbook lock for some time, I can move the new recently cancelled map to OrdermatchContext instead with its own lock to reduce the delay.

Sorry about more confusion I am about to create @mariocynicys @onur-ozkan but I will move recently cancelled map back to orderbook struct and handle it in both `insert_or_update_order_update_trie` and `delete_order` where it's already locked. I was reluctant to do that at first since recently cancelled will be checked when order is being added by means other than maker order created message but once cancel order message is broadcasted an order shouldn't be added by any means whether through creation message or pubkey sync.

shamardy commented:
Reverted the moving of recently cancelled to it's own lock here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/e0e5d41a1f129213a7dab0dc65e9adfb7e30a075 and fixed this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#issuecomment-2387723871 here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/cdfd809e2db4f8187c5d8a10f58ade118831f459.

shamardy commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -331,6 +332,30 @@ async fn process_orders_keep_alive(
+        let ordermatch_ctx = OrdermatchContext::from_ctx(ctx).expect("from_ctx failed");
+        let recently_cancelled = ordermatch_ctx.recently_cancelled.lock();
+        if let Some(order_pubkey) = recently_cancelled.get(&uuid) {
+            if order_pubkey == &from_pubkey {
```
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/commit/cdfd809e2db4f8187c5d8a10f58ade118831f459

mariocynicys commented on mm2src/coins/eth/web3_transport/websocket_transport.rs:
```diff
@@ -206,9 +198,6 @@ impl WebsocketTransport {
-
                         let request_id = id.as_u64().unwrap_or_default() as usize;
 
                         if let Some(notifier) = response_notifiers.remove(&request_id) {
```
since `remove` doesn't clear entries now. such a pre-clear is needed for correctness.

shamardy replied:
It seems that it will be better to check expiry on remove as well and leave `clear_expired_entries` as private.

shamardy replied:
Done https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/abfb252ede09f7aa04926627e751363635f90f48

shamardy commented on mm2src/mm2_main/tests/mm2_tests/mm2_tests_inner.rs:
```diff
@@ -1564,6 +1564,9 @@ fn test_cancel_order() {
+/// in this case Alice gets the order creation message from p2p cache using IHAVE / IWANT control messages.
 #[test]
 #[cfg(not(target_arch = "wasm32"))]
 fn test_cancel_all_orders() {
```
> I think we should **intentionally** send the order cancellation message before creating it.

There is no way to do that intentionally since to cancel an order from the maker node it has to have the order created first, in the current test order cancellation almost always arrive first, I never found a case where it didn't but changed the check for log that the creation message was ignored in case it doesn't happen. But I will try to send the order cancellation message first intentionally in a new test using `maker_order_cancelled_p2p_notify` straight away from ctx, if I succeed will revert the changes to the current test.

shamardy replied:
> But I will try to send the order cancellation message first intentionally in a new test using maker_order_cancelled_p2p_notify straight away from ctx, if I succeed, I will revert the changes to the current test.

Couldn't do it through ctx, but added a test here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/598045dbb779a8543294b7183db73ab316e345a7 that is deterministic and cancel message is always received before creation.

shamardy commented:
Now `test_cancel_all_orders` fails after the latest additions 😩

shamardy commented:
> Now `test_cancel_all_orders` fails after the latest additions

I understand why, will fix it. The below claim https://github.com/KomodoPlatform/komodo-defi-framework/blob/860544f1b431a58d6fce20e4386efb3bc3f4245c/mm2src/mm2_main/src/lp_ordermatch.rs#L380-L381 is not right as we can receive it through orderbook call and receive order creation message later.

shamardy commented on mm2src/mm2_core/src/data_asker.rs:
```diff
@@ -113,8 +113,6 @@ pub async fn send_asked_data_rpc(
     let mut awaiting_asks = ctx.data_asker.awaiting_asks.lock().await;
-    awaiting_asks.clear_expired_entries();
-
     match awaiting_asks.remove(&asked_data.data_id) {
```
Fixed by checking expiry on remove https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/abfb252ede09f7aa04926627e751363635f90f48

shamardy commented on mm2src/common/expirable_map.rs:
```diff
@@ -19,48 +19,90 @@ impl<V> ExpirableEntry<V> {
+    /// Iterates through the `expiries` in order, removing entries that have expired.
+    /// Stops at the first non-expired entry, leveraging the sorted nature of `BTreeMap`.
+    ///
+    /// # Implementation notes:
+    /// Why `pop_first()` usage is better than `first_key_value()` in this case:
+    ///
+    /// - `pop_first()`: Removes expired entries efficiently without an extra remove operation.
+    /// - `first_key_value()`: Wouldn't require re-insertion for non-expired entries,
+    ///   but `pop_first()` only needs to do this once per function call.
```
> also it looks like, based on that doc, that `first_key_value` is better!

No as it requires extra remove operation on each expired entry.

shamardy replied:
> ummmm, this wouldn't be an interesting detail when LSP shows it up. let's move it as a comment inside?

Moved it inside here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/5a839be58c7b5e4338024c5b0df561eaaeaea428

onur-ozkan commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -331,6 +332,17 @@ async fn process_orders_keep_alive(
     Ok(())
 }
 
+fn process_maker_order_created(
+    ctx: &MmArc,
+    from_pubkey: String,
+    created_msg: new_protocol::MakerOrderCreated,
+) -> OrderbookP2PHandlerResult {
+    let order: OrderbookItem = (created_msg, from_pubkey).into();
+    insert_or_update_order(ctx, order);
+
+    Ok(())
```
This seems like a error-free function. We can remove the return type.

shamardy replied:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/158252c6d60ddf24d7dea72abe43f8a5c7976a99

onur-ozkan commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -350,6 +362,28 @@ fn process_maker_order_updated(
     Ok(())
 }
 
+fn process_maker_order_cancelled(
+    ctx: &MmArc,
+    from_pubkey: String,
+    cancelled_msg: new_protocol::MakerOrderCancelled,
+) -> OrderbookP2PHandlerResult {
+    let uuid = Uuid::from(cancelled_msg.uuid);
+    let ordermatch_ctx = OrdermatchContext::from_ctx(ctx).expect("from_ctx failed");
+    let mut orderbook = ordermatch_ctx.orderbook.lock();
+    // Add the order to the recently cancelled list to ignore it if a new order with the same uuid
+    // is received within the `RECENTLY_CANCELLED_TIMEOUT` timeframe.
+    // We do this even if the order is in the order_set, because it could have been added through
+    // means other than the order creation message.
+    orderbook.recently_cancelled.insert(uuid, from_pubkey.clone());
+    if let Some(order) = orderbook.order_set.get(&uuid) {
+        if order.pubkey == from_pubkey {
+            orderbook.remove_order_trie_update(uuid);
+        }
+    }
+
+    Ok(())
+}
```
Same here.

shamardy replied:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/158252c6d60ddf24d7dea72abe43f8a5c7976a99

onur-ozkan commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2490,12 +2510,34 @@ struct Orderbook {
 
+impl Default for Orderbook {
+    fn default() -> Self {
+        Orderbook {
+            ordered: HashMap::default(),
+            pairs_existing_for_base: HashMap::default(),
+            pairs_existing_for_rel: HashMap::default(),
+            unordered: HashMap::default(),
+            order_set: HashMap::default(),
+            pubkeys_state: HashMap::default(),
+            recently_cancelled: TimeCache::new(RECENTLY_CANCELLED_TIMEOUT),
+            topics_subscribed_to: HashMap::default(),
+            memory_db: MemoryDB::default(),
+            my_p2p_pubkeys: HashSet::default(),
+        }
```
minor suggestion:

```suggestion
        Orderbook {
            recently_cancelled: TimeCache::new(RECENTLY_CANCELLED_TIMEOUT),
            ..Default::default()
        }
```

onur-ozkan replied:
Clippy is being angry on this. No idea why..

shamardy replied:
Yeah, just faced it too :)

shamardy replied:
Makes sense, this pattern should be used in new/constructor implementation as it resolves the rest of the fields to default from the default implementation, but we try it here inside default implementation.

onur-ozkan reviewed (dismissed):
Final notes from my side, LGTM other than these notes:

mariocynicys commented on mm2src/mm2_main/tests/mm2_tests/orderbook_sync_tests.rs:
```diff
@@ -1308,6 +1308,79 @@ fn setprice_min_volume_should_be_displayed_in_orderbook() {
+    }))
+    .unwrap();
+
+    // Check Alice's orderbook again to make sure the order was re-inserted after the cancellation.
```
`wasn't re-inserted after cancellation`*

shamardy replied:
Fixed, copy paste problems :)

mariocynicys commented on mm2src/mm2_main/tests/mm2_tests/orderbook_sync_tests.rs:
```diff
@@ -1308,6 +1308,79 @@ fn setprice_min_volume_should_be_displayed_in_orderbook() {
+    // Check Alice's orderbook again to make sure the order was re-inserted after the cancellation.
+    let orderbook = block_on(orderbook_v2(&mm_alice, "RICK", "MORTY"));
+    let asks = orderbook["result"]["asks"].as_array().unwrap();
+    // Alice should see the order in the orderbook as she got it through `GetOrderbook` p2p request.
+    assert_eq!(asks.len(), 0, "Alice RICK/MORTY orderbook must have exactly 1 ask");
```
`shouldn't see the order`*, and also the assert message should be `0` asks not `1`

shamardy replied:
Fixed

mariocynicys reviewed (commented):
Last points, LGTM otherwise :)

mariocynicys commented on mm2src/mm2_core/Cargo.toml:
```diff
@@ -20,6 +20,7 @@ lazy_static = "1.4"
 mm2_err_handle = { path = "../mm2_err_handle" }
 mm2_event_stream = { path = "../mm2_event_stream" }
 mm2_metrics = { path = "../mm2_metrics" }
+mm2-libp2p = { path = "../mm2_p2p", package = "mm2_p2p" }
```
nit `mm2-libp2p` -> `mm2_libp2p` for consistency

mariocynicys reviewed (approved):
LGTM
